include/ginkgo/core/base/utils.hpp

```diff
@@ -325,6 +325,32 @@ inline const typename std::decay<T>::type *as(const U *obj)
 }
 
 
+/**
+ * Performs polymorphic type conversion of a unique_ptr.
+ *
+ * @tparam T  requested result type
+ * @tparam U  static type of the passed object
+ *
+ * @param obj  the unique_ptr to the object which should be converted.
+ *             If successful, it will be reset to a nullptr.
+ *
+ * @return If successful, returns a unique_ptr to the subtype, otherwise throws
+ *         NotSupported.
+ */
+template <typename T, typename U>
+inline std::unique_ptr<typename std::decay<T>::type> as(
+    std::unique_ptr<U> &&obj)
+{
+    if (auto p = dynamic_cast<typename std::decay<T>::type *>(obj.get())) {
+        obj.release();
+        return std::unique_ptr<typename std::decay<T>::type>{p};
+    } else {
+        throw NotSupported(__FILE__, __LINE__, __func__,
+                           name_demangling::get_type_name(typeid(obj.get())));
+    }
+}
+
+
```

Comment:
At least for uniformity, does it not make sense to also implement a `std::shared_ptr` version?

Reply:
That's true, so you mean for a shared_ptr const/non-const lvalue?

Reply:
I agree, I think it'd be really nice to have `as` work for basically any type of smart or plain pointer.

Reply:
For `std::shared_ptr` I guess both const/non-const can be supported, at least with `dynamic_pointer_cast` for example.

Reply:
Yes, I decided to add them both, since we often use shared_ptr<const ...> in our code.